### DISCUSSION TITLE
How to use Kedro for Lakehouse ETL ?

### DISCUSSION THREAD

**Discussion by nthanhtung** [kedro-org/kedro#1009] [Idea]
**_How to use Kedro for Lakehouse ETL ?_**

# Lakehouse Architecture
* Lakehouse = data lake + data warehouse
* Providers: Databricks (Multi Cloud) , Synapse Workspace (Only Azure)
https://www.youtube.com/watch?v=d7Px_MZV05I

# Requirement for ETL Tool
**_Easy for traditional database developer, citizen developer to adopt_**
* Support SQL: easy to learn/code/maintain/migrate from legacy code
* Change Data Capture for incremental load
* Schema drift, Data Quality
* Linage graph for object dependency
* Parameterized, reusable objects to avoid duplicate code 
* Alert, Monitoring, Log
* Failover, Retry
* Managed service: avoid headache for VM, Docker
* No code, low code UIUX: setup environment, scale up/out workload, DevOps
* Security: Private Vnet, MFA linked with organization account
* Easy to Integrate with other components in Data Platform to build end to end data pipeline in 1 tool (event hub trigger -> process data using spark -> train ML model -> generate result -> trigger Power BI refresh -> send alert Slack to user)

# Available ETL Tool that can support the above requirement
## Azure Data Factory 
* Code free ETL as a service: https://www.youtube.com/watch?v=CZQOxPY7UuA&t=440s
## Databricks Delta 
* SQL support ACID for data warehouse : https://docs.databricks.com/delta/index.html  
* Delta live table: https://www.youtube.com/watch?v=6Q8qPZ7c1O0&t=326s  | https://www.youtube.com/watch?v=cjXkhmxi-aY&t=858s
* Auto loader: https://www.youtube.com/watch?v=kNG-5uRY4o8&t=474s
* Multi job dependency management: https://www.youtube.com/watch?v=HNqPtUQV534 , https://www.youtube.com/watch?v=1AotjGNLo3Q&t=1232s

**Comment by datajoely** [comment]
Hi @nthanhtung you ask a very good question - something which truly makes us ask what Kedro should and shouldn't be. As the product has evolved, my view is that it is best described as a 'ML engineering framework'. 

What does this mean in practice? Kedro provides a mechanism for organising your Python ML process in a structured, battle-tested pattern in a data-centric way of working with batch data.

Going through your requirements, I've gone through how I think Kedro fits into the world you describe:

| # | Point | Comment |
|---|---|---|
| 1 | Support SQL: easy to learn/code/maintain/migrate from legacy code  | Sort of - but only to pull data outside of SQL into the Python (or Python API) world. |
| 2 | Change Data Capture for incremental load | Somewhat, via the `ParittionedDataSet` and `IncrementalDataSet` |
| 3 | Schema drift, Data Quality | Not out of the box, via custom hooks |
| 4 | Lineage graph for object dependency | Yes, 1st level (table level) not 2nd level (schema level). |
| 5 | Parameterized, reusable objects to avoid duplicate code | Yes, watch this space for how `Modular Pipelines` and micropackaging evolve |
| 6 | Alert, Monitoring, Log | Not out of the box, possible via hooks |
| 7 | Failover, Retry | Not out of the box, possible via hooks |
| 8 | Managed service: avoid headache for VM, Docker | Nope - It's a development framework, we want to work within the managed services |
| 9 | No code, low code UIUX: setup environment, scale up/out workload, DevOps | Kedro doesn't concern itself with infra. This topic can also mean many different things to different people. |
| 10 | Security: Private Vnet, MFA linked with organization account | See points 8,9 |
| 11 | Easy to Integrate with other components in Data Platform to build end to end data pipeline in 1 tool (event hub trigger -> process data using spark -> train ML model -> generate result -> trigger Power BI refresh -> send alert Slack to user) | Kedro wants to own the hidden complexity between the "process data using spark -> train ML model -> generate result" stages. From a scheduler point of view this could be 1 node, from a Kedro point of view this could be very complex at a more micro level. |

To summarise I think point 11 is the most important here - there is a lot of complexity in the middle part of that process that at a macro level is just 'a part of the ETL process', but something Kedro tries to focus on. 

A mature system could look something like this:

- Scheduler (airflow/prefect/dagster etc.)
- Platform / Execution engine (Snowflake/Databricks delta/Big Query) etc.
- SQL based Data engineering - Developed in something like [dbt](https://getdbt.com/)
- ML components - Pipelines developed in Kedro exposed as a API endpoints triggered by the scheduler.
- Data reliability / observability - Provided by tools like [Monte Carlo](https://www.montecarlodata.com/)
- If you're really sophisticated, use tools like [gantry.io](https://gantry.io/) and [fiddler.ai](https://www.fiddler.ai/) to manage the lifecycle of models.

**Reply by nthanhtung** [reply]
Hi @datajoely 

Thanks for you detail response, highly appreciate !!!
Here are my additional questions:
1. Kedro wants to own the hidden complexity between the "process data using spark -> train ML model -> generate result" stages. 

- for "process data using spark", 
- Using **Spark on Databricks** with native features including SQL notebook, Delta Live Table ETL, Auto Loader, Dependency have **already cover all use case of Lakehouse ETL data processing using spark**
- -> then I assume **Kedro** will be more suitable for **non Databricks, non Lakehouse ETL data processing using spark** use case right

2. A mature system could look something like this  ...[6 tools] ... | 

- I prefer a **single platform that has all native built-in enterprise tools for Data Engineering**, 
- No need to add any 3rd tools in the architecture to **minimize complexity for citizen developer while maximize performance/price** 
- ->Example: Synapse Workspace/Databricks

**Reply by datajoely** [reply]
## Point 1:

- Kedro is platform agnostic, so you should be able to use Kedro within a Sagemaker, Databricks or a Raspberry PI. 
- Spark means many things to many people - it's an execution engine, a DataFrame API, a SQL warehouse as well as this new hybrid Lakehouse idea. 
- Kedro has a strong philosophy that production code should not be written or managed in notebooks - but in this world there are many ways that Kedro can fit into the Databricks notebook world or just using it as a Spark back-end. 
    - You can package a Kedro [project](https://kedro.readthedocs.io/en/stable/03_tutorial/05_package_a_project.html) or specific [pipeline](https://kedro.readthedocs.io/en/stable/06_nodes_and_pipelines/03_modular_pipelines.html) and use it within a Databricks notebook. The modular pipeline packaging workflow in particular is our way of reducing re-invention across and organisation and allowing teams to share complex pieces of work.
    - You can use [Kedro's DataCatalog](https://kedro.readthedocs.io/en/stable/05_data/01_data_catalog.html) to manage your dataset references declaratively.
    - You can use the [Kedro IPython magic](https://kedro.readthedocs.io/en/latest/11_tools_integration/02_ipython.html) to get access to catalog and context dynamically for within a notebook.

## Point 2:

- So this is a valid point and fundamentally comes down to best in class vs general purpose. The stuff the Monte Carlo folks are doing is a good example of something that Databricks doesn't do yet... but may catch up eventually. Vendor lock in is also a concern for some users.
- I suppose it's also worth noting that some of the tools I mentioned aren't always substitutes for Databricks functionality. Databricks in many ways has to design for the most general of purposes. Both `dbt` and `kedro` for example have opinionated ways on how power-users can write  SQL and spark / python  respectively and simply treat Databricks as an execution target.

The other thing work looking is at is our current work on our [spark.DeltaTable](https://github.com/quantumblacklabs/kedro/pull/964) PR which includes an honest discussion about which parts fit into Kedro's philosophy and which don't.

**Reply by nthanhtung** [reply]
@datajoely 

# A specific use case: enterprise lakehouse data processing using spark in production
Data layer in Databricks Lakehouse: Bronze -> Silver -> Gold : https://docs.microsoft.com/en-us/azure/architecture/solution-ideas/articles/azure-databricks-modern-analytics-architecture

| Layer   |      Native Data Processing Tool with no/low code UIUX      |
|----------|:-------------:|
| Bronze: Ingest 500 ERP tables, 100 Kafka topics to staging tables |  Auto Loader & Azure Data Factory | 
| Silver: clean, deduplicate data in staging and merge into foundation tables |  Notebook & delta live table ETL   |  
| Gold: calculate metrics, extract features and merge/overwrite into use case tables for ML, BI | Notebook & delta live table ETL |

Support tool:
- Data Governance: Full Auto Data Linage , Glossary, Sensitive Data Tracking: Azure Purview https://azure.microsoft.com/en-us/services/purview/#overview
- CICD : Azure DevOps

1. Key features of Native Data Processing Tool : Point 1 -> 7 in the above reply
2. **_In this specific use case_, What features of Kedro has/is better, that native tools don't have, that I should convince my client to use.** 
3. **And if I use that features, how many working hours I can save to build a pipeline and deploy to production**
4. **And how much money spending for computing power I could save by gaining how much performance?**
5. **_Kedro has a strong philosophy that production code should not be written or managed in notebooks_** -> but Databricks notebook has many rich features vs Jupiter Notebook (point 1 -> 7), and I did see many folks in large tech savvy enterprise use Databricks notebook  to process data for production workload


# Databricks in many ways has to design for the most general of purposes -> Not really
In Databricks website, they position themself as: "All your data, analytics and AI on one lakehouse platform"
They are not only target execution engine, They continuously introduce specific features to serve specific use cases:
 1. Delta table with ACID, photon engine, SQL endpoint for enterprise graded data warehouse on data lake, even with row-level security defined by only using SQL
 2. Delta live table ETL for data quality control, data linage, ETL workflow -> similar features to this Monte Carlo demo clip:  https://www.youtube.com/watch?v=sMjmzpwPIhU&t=1520s
 3. Auto loader for streaming change data capture automatically
 4. Even create dashboard visualization directly in databricks
 5. ....etc....

Same for Synapse Workspace: build specific features, combine all data products including Spark, Data Factory, SQL DB, Power BI, no-code Auto ML, data governance into 1 place
**_Databricks/Microsoft both have open architecture for external tool, given the native tools is not sufficient for the use cases_**

# Vendor lock issue -> Not really
"Vendor lock-in is when someone is essentially forced to continue using a product or service regardless of quality, because switching away from that product or service is not practical."
"In economics, vendor lock-in, also known as proprietary lock-in or customer lock-in, makes a customer dependent on a vendor for products and services, unable to use another vendor without substantial switching costs."

Given data platform vendor = Databricks/Microsoft

## For now
1. If there's better lakehouse platform than Databricks/Synapse, better ETL tool than Azure Data Factory, Delta live table, 
2. and the new tool is proven to be time saving, cost saving
3. I will still be able to migrate from old tool to new tool

## In the future 
1. Assumption: Enterprise become too dependent on single data platform from single vendor
2. The vendor no longer support products -> M$ have good revenue stream from this product -> unlikely to happen
3. The vendor no longer have great engineer to innovate product -> M$ has lots of cash to sponsor for R&D -> unlikely to happen
4. Given 2 = true or 3 = true -> Other players like AWS/GCP are very aggressive to gain revenue, and build better product -> they will help you to migrate away from M$

**Reply by datajoely** [reply]
I'm really sorry - but I'm not sure what you're asking at this point? Databricks and Kedro are trying to solve different problems, one very broad one very narrow.

**Reply by nthanhtung** [reply]
Hi, 
In this specific use case about spark data processing
What features of Kedro has/is better, that native tools don't have, and the benefits in time saving / cost saving





> @datajoely
> 
> # A specific use case: enterprise lakehouse data processing using spark in production
> Data layer in Databricks Lakehouse: Bronze -> Silver -> Gold : https://docs.microsoft.com/en-us/azure/architecture/solution-ideas/articles/azure-databricks-modern-analytics-architecture
> 


| Layer   |      Native Data Processing Tool with no/low code UIUX      |
|----------|:-------------:|
| Bronze: Ingest 500 ERP tables, 100 Kafka topics to staging tables |  Auto Loader & Azure Data Factory | 
| Silver: clean, deduplicate data in staging and merge into foundation tables |  Notebook & delta live table ETL   |  
| Gold: calculate metrics, extract features and merge/overwrite into use case tables for ML, BI | Notebook & delta live table ETL |

> Support tool:
> 
> * Data Governance: Full Auto Data Linage , Glossary, Sensitive Data Tracking: Azure Purview https://azure.microsoft.com/en-us/services/purview/#overview
> * CICD : Azure DevOps

**Reply by datajoely** [reply]
So Kedro is a data-centric development workflow - Databricks is the execution target/storage layer among other things

**Reply by nthanhtung** [reply]
for data-centric development workflow, there are several cloud-based managed service (no docker, VM), no/low code UI, native tools:
* Databricks delta live table ETL 
* Multiple interdependent Databricks Notebooks orchestrated by ADF, or Databricks Jobs
* Multiple interdependent Synapse Notebooks/ Store Procedures orchestrated by ADF
* ADF wrangling data flow with spark
